server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -386,24 +400,31 @@ static void parseObjectOrNested(ParseContext context, ObjectMapper mapper) throw
 
     private static void innerParseObject(ParseContext context, ObjectMapper mapper, XContentParser parser,
                                          String currentFieldName, XContentParser.Token token) throws IOException {
+        assert token == XContentParser.Token.FIELD_NAME || token == XContentParser.Token.END_OBJECT;
+        String[] paths = null;
         while (token != XContentParser.Token.END_OBJECT) {
-            if (token == XContentParser.Token.START_OBJECT) {
-                parseObject(context, mapper, currentFieldName);
-            } else if (token == XContentParser.Token.START_ARRAY) {
-                parseArray(context, mapper, currentFieldName);
-            } else if (token == XContentParser.Token.FIELD_NAME) {
+            if (token == XContentParser.Token.FIELD_NAME) {
                 currentFieldName = parser.currentName();
+                paths = splitAndValidatePath(currentFieldName);
                 if (MapperService.isMetadataField(context.path().pathAsText(currentFieldName))) {
                     throw new MapperParsingException("Field [" + currentFieldName + "] is a metadata field and cannot be added inside"
                         + " a document. Use the index API request parameters.");
                 }
+            } else if (isDisabled(mapper, paths)) {
```

Comment:
This existing method is structured in a fairly confusing way, but given the way it's structured, I think it'd be clearer to check this as part of the 'field name' block: 

```
if (token == XContentParser.Token.FIELD_NAME) {
    currentFieldName = parser.currentName();
    paths = splitAndValidatePath(currentFieldName);
    if (MapperService.isMetadataField(context.path().pathAsText(currentFieldName))) {
        throw new MapperParsingException("Field [" + currentFieldName + "] is a metadata field and cannot be added inside"
            + " a document. Use the index API request parameters.");
        } else if (isDisabled(mapper, paths)) {
            parser.nextToken();
            parser.skipChildren();
        }
    }
} else if ...
```